providers/defltprov.c

```diff
@@ -380,6 +385,9 @@ static const OSSL_ALGORITHM deflt_keymgmt[] = {
 #endif
 #ifndef OPENSSL_NO_DSA
     { "DSA", "default=yes", dsa_keymgmt_functions },
+#endif
+#ifndef OPENSSL_NO_EC
+    { "EC:ECDH", "default=yes", ec_keymgmt_functions },
```

Comment:
Ah ok! Is it documented somewhere? I was looking for this bit of information in `internal/man7` but failed to find it.

Reply:
Also this does not seem to hold for DSA or DH keys.

Would it be equally fine to have a single keymgmt entry look like this?

```C
    { "EC:id-ecPublicKey", "default=yes", ecdh_keymgmt_functions },
```

while for ECDH and ECDSA the entries would look like this:

```C
    { "ECDH:id-ecPublicKey", "default=yes", ecdh_keyexch_functions },
```

```C
    { "ECDSA:id-ecPublicKey", "default=yes", ecdsa_signature_functions },
```

Reply:
The algorithm names you use to look up the key exchange algorithm must have equivalent algorithm names in key management. So if you look up a key exchange algorithm via the name "ECDH", then you expect to find an "ECDH" key management algorithm. Similarly for signature. Therefore a single line would need to look like this:

````
    { "ECDH:ECDSA:id-ecPublicKey", "default=yes", ec_keymgmt_functions },
````

In fact, thinking about it, it probably *has* to be on a single line rather than two separate ones because otherwise there is ambiguity with the "id-ecPublicKey" name.

> I was looking for this bit of information in internal/man7 but failed to find it.

It should be there but I'm failing to find a good reference. It is stated explicitly in provider-keyexch.pod that every key exchange algorithm must have an equivalent key management algorithm:

"A key exchange algorithm must also implement some mechanism for generating,
loading or importing keys via the key management (OSSL_OP_KEYMGMT) operation.
See L<provider-keymgmt(7)> for further details."

There are similar words for provider-signature.pod. But it doesn't explicitly state that the algorithms names need to be the same.

Reply:
With a single line like that, what you end up doing is to make ECDH and ECDSA aliases of each other. I am pretty sure that's *not* what you want. Remember that all of the EVP stuff share *one* namemap.

We had a related discussion when disputing RSA vs RSA-PSS, remember that? I was talking about algorithms being able to specify what key type they need, by name. That would be quite suitable here, i.e. the ECDSA and ECDH implementations should somehow be able to say that they uses EC keys. Incidently, this corresponds to how this is implemented in libcrypto (something Stephen worked *hard* on, 'cause we actually had separate ECDH and ECDSA key  types originally)

Reply:
> With a single line like that, what you end up doing is to make ECDH and ECDSA aliases of each other. I am pretty sure that's _not_ what you want. Remember that all of the EVP stuff share _one_ namemap.
> 
> We had a related discussion when disputing RSA vs RSA-PSS, remember that? I was talking about algorithms being able to specify what key type they need, by name. That would be quite suitable here, i.e. the ECDSA and ECDH implementations should somehow be able to say that they uses EC keys. Incidently, this corresponds to how this is implemented in libcrypto (something Stephen worked _hard_ on, 'cause we actually had separate ECDH and ECDSA key types originally)

I am a bit confused, wouldn't then a collision happen anyway because of `"id-ecPublicKey"` once we have 

```C
    { "ECDH:id-ecPublicKey", "default=yes", ecdh_keyexch_functions },
```

and

```C
    { "ECDSA:id-ecPublicKey", "default=yes", ecdsa_signature_functions },
```

Reply:
Yes it would.  Those lines are wrong anyway, 'cause there's *nothing* in "id-ecPublicKey" that says that this is a DH or DSA implementation.

A reason for confusion is when you go from `crypto/ec/ec_pmeth.c` while thinking in ECDH and ECDSA terms, 'cause the EVP_PKEY_METHOD defined there is in terms of a EC key, not a ECDH or ECDSA key, and more or less assumes that at least the signature function actually computes ECDSA, whooptidoo (it's very visible when you study `crypto/ec/ecdsa_sign.c`, which is basically a set of calls into an EVP_PKEY_METHOD structure).

The naïve way to go could be to simply have a EC implementation that simply does the right thing given...  something, I dunno what.  But that would probably be problematic as soon as we start hitting variants like X25519 or EdDSAwhatnot etc, hence "naïve".

The other way would be for certain algorithm to specify that they want a key of some specific type name that is *not* the name of the algorithm, so an "ECDSA" implementation could say that it wants a "EC" key.  The natural way to do so would be in form of a parameter.

Reply:
Since the keymgmt implementation *must* exist in the same provider as the algorithm, type name string matching shouldn't be a problem, as a provider is presumably authored by the same person or same  group.

Reply:
Thoughts?  I'm ready to implement the necessary infrastructure to allow implementations to specify a key type name.

Reply:
@levitte 
> The other way would be for certain algorithm to specify that they want a key of some specific type name that is not the name of the algorithm, so an "ECDSA" implementation could say that it wants a "EC" key. The natural way to do so would be in form of a parameter.

Uhm, I think the latter is probably needed in this case.

- ECDH and ECDSA share the exact same kind if keys as we are discussing;
- SM2 is a different signature scheme that uses EC keys (well, ok, _almost_, as they are identical but in the keygen operation as the range of valid secret scalars is different, so probably for SM2 we should end up with a separate, almost identical, dedicated keymgmt)
- But definitely ECDSA is not the only signature scheme that can be built on top of EC

Reply:
@levitte

> Yes it would. Those lines are wrong anyway, 'cause there's nothing in "id-ecPublicKey" that says that this is a DH or DSA implementation.

Yet, if I don't add "id-ecPublicKey" as an alias, the interaction between PEM parser and EVP does not lead currently to a match with the namemap.
Finding out about "id-ecPublicKey" in the debugger was the cause for my gratuitous rant about our extensive use of macros for defining OSSL_store_* and namemap functions.

Reply:
@levitte

What about an alternative version of @mattcaswell first comment?

```C
    { "ECDH", "default=yes", ec_keymgmt_functions },
    { "ECDSA", "default=yes", ec_keymgmt_functions },
    { "SM2", "default=yes", sm2_keymgmt_functions },

    { "ECDH", "default=yes", ecdh_keyexch_functions },

    { "ECDSA", "default=yes", ecdsa_signature_functions },
    { "SM2", "default=yes", sm2_signature_functions },
```

But then how does the PEM/EVP layer match a decoded EC key with the provided ECDH keyexch or ECDSA signature ?

Reply:
> But then how does the PEM/EVP layer match a decoded EC key with the provided ECDH keyexch or ECDSA signature ?

That's the hard part, since the key type name doesn't match the algorithm name.

I don't like the alternative version at all.  Do we really want to go back to having ECDH and ECDSA keys?  ["Here be dragons"](https://en.wikipedia.org/wiki/Here_be_dragons) feels like a very appropriate thing to say here.

As for SM2 keys, considering they are generated differently enough to make them a different key type, at least with rgards to keymgmt.  Serializers will have to adapt (which shouldn't even be hard).

(Incidently, I just discovered that we have a "ECDSA PUBLIC KEY" PEM name...  but then found that we don't actually implement it, so I assume it's an old remnant that we should remove)

Reply:
> I don't like the alternative version at all. Do we really want to go back to having ECDH and ECDSA keys? "Here be dragons" feels like a very appropriate thing to say here.

I agree it would be neater if we had a way to say "here you are, this is bunch of keyexch and signature algorithms this provider offers, they all work on key objects that are manipulated through this keymgmt algorithm"

Still, it does not make it easy for PEM/EVP, in the presence of a provider offering e.g. 2 keyexch algorithms using EC keys and 4 signature algorithms using the same kind of key objects to decide which algorithm to fetch upon EVP_PKEY_derive or EVP_PKEY_sign.

What is the plan there?

Reply:
PEM shouldn't have to care about exact algorithms, as it deals with *keys* (and other type of data, but that's the point, it's about *data types*, not what you do with them).

> What is the plan there?

Plan?  Haven't come that far yet, I'm currently formulating ideas
(you ok, sorry, I'm mincing words :wink:)

The *idea* I have in my mind is that we could have a parameter, say "keytype", that algorithm can answer to *as an option*.  If they don't answer to that parameter, libcrypto can be free to assume that the keytype, if one is needed, has a name that matches the algorithm name (which is the current assumption, i.e. it's backward compatible with what we do right now).

So back to actual planing, it looks like the infrastructure for this should be implemented ASAP.
How much patience do you (or we) have?

Reply:
Actually, it turns out that my idea is fairly simple to implement.  All that's needed is to get the param, and if the algo answers to it, find the namemap number and use that, otherwise the namemap number of the algorithm that's being used.

Reply:
Well, I can tackle the other issues here in the meantime and keep the current entries (or if you have a better interim suggestion) until we have a better solution, so that I can still benefit from evp_test and apps for testing and debugging.

Priority wise I have been holding this back for quite some time because of workload, so my main worry is that others (@InfoHunter for example has SM2 tasks pending) are waiting for this to get in to start working on other depending tasks.

The curve25519 and curve448 tasks are somewhat lateral as they have their own separate algorithms, keymgmt and encoding/decoding so if I really don't have anything else to do I could start working on those transitions. 

So in the end I am probably the wrong one to give an answer on how much patience do we have :P

Reply:
> The curve25519 and curve448 tasks are somewhat lateral as they have their own separate algorithms, keymgmt and encoding/decoding so if I really don't have anything else to do I could start working on those transitions.

Sounds like a good distraction while waiting for me :wink:

Reply:
So, er, #10636

Reply:
Another very obvious variant is to assume that the "EC" algorithm does DH for key exchange and DSA for signature, and have these:

``` C
static const OSSL_ALGORITHM deflt_keyexch[] = {
...
    { "EC:id-ecPublicKey", "default=yes", ecdh_keyexch_functions },
...
};
...
static const OSSL_ALGORITHM deflt_keymgmt[] = {
...
    { "EC", "default=yes", ec_keymgmt_functions },
...
};
```

Reply:
I don't like this particularly: I don't foresee alternative key exchange methods for EC, but who knows, back when X25519 was implemented inside the EC module as EC_CUSTOM stuff we had an alternative way of doing key exchange that was not ECDH as it is standardized (although mathematically is arguably equivalent).

And the trick would not work for ECDSA anyway, as alternative signature schemes already exist.

Reply:
Ok then.  In that case, I think #10647, coupled with a project regime to have distinct key types for different forms of EC keys (so SM2 would be a distinct type, based on specific key parameters), is the answer.